src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,255 @@
+        val bases = new Array[Byte](len)
+        val quals = new Array[Byte](len)
+
+        forloop(from=0, until=len) { i =>
+          val aBase = a.bases(i)
+          val bBase = b.bases(i)
+          val aQual = a.quals(i)
+          val bQual = b.quals(i)
+
+          val (base, qual) = {
+            if      (aBase == NoCall || bBase == NoCall) (NoCall, PhredScore.MinValue)
+            else if (aBase == bBase) (aBase, (aQual + bQual).toByte)
+            else if (aQual > bQual)  (aBase, (aQual - bQual).toByte)
+            else if (bQual > aQual)  (bBase, (bQual - aQual).toByte)
+            else (NoCall, PhredScore.MinValue)
+          }
+
+          bases(i) = base
+          quals(i) = qual
+        }
+
+        Some(DuplexConsensusRead(id=id, bases, quals, a.truncate(bases.length), b.truncate(bases.length)))
+      case _ =>
+        None
+    }
+  }
+
+  /**
+    * Creates a SAMRecord with a ton of additional tags annotating the duplex read.
+    */
+  override protected def createSamRecord(read: DuplexConsensusRead, readType: ReadType): SAMRecord = {
+    val r = super.createSamRecord(read, readType)
+    val ab = read.abConsensus
+    val ba = read.baConsensus
+
+    // Calculate the total depths and errors across both SS consensus reads
+    val totalDepths = read.abConsensus.depths.zip(read.baConsensus.depths).map(x => x._1 + x._2)
+    val totalErrors = new Array[Short](totalDepths.length)
+    forloop (from=0, until=totalErrors.length) { i =>
+      var e = 0
+      if (read.bases(i) == ab.bases(i)) e += ab.errors(i) else e += ab.depths(i) - ab.errors(i)
+      if (read.bases(i) == ba.bases(i)) e += ba.errors(i) else e += ba.depths(i) - ba.errors(i)
+      totalErrors(i) = e.toShort
+    }
+
+    { import ConsensusTags.PerRead._
+      r.setAttribute(RawReadCount,       totalDepths.max.toInt)
+      r.setAttribute(MinRawReadCount,    totalDepths.min.toInt)
+      r.setAttribute(RawReadErrorRate,   totalErrors.sum / totalDepths.sum.toFloat)
+      r.setAttribute(AbRawReadCount,     read.abConsensus.depths.max.toInt)
+      r.setAttribute(BaRawReadCount,     read.baConsensus.depths.max.toInt)
+      r.setAttribute(AbMinRawReadCount,  read.abConsensus.depths.min.toInt)
+      r.setAttribute(BaMinRawReadCount,  read.baConsensus.depths.min.toInt)
+      r.setAttribute(AbRawReadErrorRate, read.abConsensus.errors.sum / read.abConsensus.depths.sum.toFloat)
+      r.setAttribute(BaRawReadErrorRate, read.baConsensus.errors.sum / read.baConsensus.depths.sum.toFloat)
+    }
+
+    { import ConsensusTags.PerBase._
+      // TODO: should we also set the `RawReadCount` and `RawReadErrors` tags? Totally redundant, but maybe nice for consistency?
+      r.setAttribute(AbRawReadCount, read.abConsensus.depths)
+      r.setAttribute(BaRawReadCount, read.baConsensus.depths)
+      r.setAttribute(AbRawReadErrors, read.abConsensus.errors)
+      r.setAttribute(BaRawReadErrors, read.baConsensus.errors)
+    }
+
+    r
```

Comment:
😝 Are you going to make me rename `ab` and `ba` next?  (I change r -> rec for you).